File: cpp/problems/leetcode/0055_jump_game/solution.cpp
// LeetCode 55
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n = (int)nums.size(), far = 0;
        for (int i = 0; i < n; ++i) {
            if (i > far) return false;
            far = max(far, i + nums[i]);
            if (far >= n - 1) return true;
        }
        return true;
    }
};

int main() {
    vector<int> a{2, 3, 1, 1, 4};
    assert(Solution().canJump(a));
    vector<int> b{3, 2, 1, 0, 4};
    assert(!Solution().canJump(b));
    vector<int> c{0};
    assert(Solution().canJump(c));
    cout << "0055 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0055_jump_game/solution.py
"""LeetCode 55: Jump Game."""

from __future__ import annotations


class Solution:
    def canJump(self, nums: list[int]) -> bool:
        far = 0
        n = len(nums)
        for i in range(n):
            if i > far:
                return False
            far = max(far, i + nums[i])
            if far >= n - 1:
                return True
        return True


if __name__ == "__main__":
    assert Solution().canJump([2, 3, 1, 1, 4])
    assert not Solution().canJump([3, 2, 1, 0, 4])
    assert Solution().canJump([0])
    print("0055 OK")
